test: add test case to verify concurrent reading and writing transactions always work correctly

ahrtr opened the pull request:
Linked to https://github.com/etcd-io/bbolt/issues/369

More test cases will be added in separate PR.

ahrtr commented:
Let me make the test case more generic.

cenkalti commented on concurrent_test.go:
```diff
@@ -0,0 +1,91 @@
+	// make sure all readonly transactions have already started
+	t.Log("Wait for all readonly transactions to get started.")
+	<-startCh
+	<-startCh
```
Can you use a WaitGroup here as well?

ahrtr commented:
Ready for review. cc @ptabor @serathius @cenkalti @tjungblu

cenkalti commented:
LGTM

cenkalti commented on concurrent_test.go:
```diff
@@ -0,0 +1,92 @@
+	"go.etcd.io/bbolt/internal/btesting"
+)
+
+func TestRepeatableRead(t *testing.T) {
```
Nit: Have a `const` block here to define parameters for the test to see all of them together. Example:
``` 
const (
    readerInterval = 50 * time.Millisecond
    writerInterval = 10 * time.Millisecond
    testDuration = 30 * time.Second
)
```

ahrtr replied:
done.

ahrtr commented:
cc @ptabor

ptabor commented on concurrent_test.go:
```diff
@@ -0,0 +1,98 @@
+	key := []byte("foo")
+
+	// prepare the db
+	t.Log("Create the db")
```
nit: If we want such log (I have nothing against) it should be part of `btesting.(Must)CreateDB`

ptabor commented on concurrent_test.go:
```diff
@@ -0,0 +1,98 @@
+				b := tx.Bucket(bucket)
+				v := fmt.Sprintf("%04d", rand.Intn(10000))
+				putErr := b.Put(key, []byte(v))
+				require.NoError(t, putErr)
```
Probably we should bubble the error up (see next comment)

ptabor commented on concurrent_test.go:
```diff
@@ -0,0 +1,98 @@
+
+				return nil
+			})
+			require.NoError(t, err)
```
I assume require.... is likely to call (https://pkg.go.dev/testing#T.FailNow), that shouldn't be called from other goroutines than the main tests go-routine. 

Please consider `errgroup` instead of `wg`
https://pkg.go.dev/golang.org/x/sync/errgroup

ptabor commented on concurrent_test.go:
```diff
@@ -0,0 +1,98 @@
+		defer wg.Done()
+		for {
+			select {
+			case <-stopCh:
```
optional: probably more 'modern' way of this is to use ctx.Done() on nested context.

ptabor commented on concurrent_test.go:
```diff
@@ -0,0 +1,98 @@
+	var wg sync.WaitGroup
+	t.Log("Create readonly transactions")
+	wg.Add(1)
+	go func() {
```
would be good to generalize this code to maintain multiple RO transaction started at random/the same times.

ptabor replied:
Part of this might be factoring the reading and writing workers to a separate methods such that they can be reused for similar tests - e.g working on the bigger volume of data.

ptabor commented on concurrent_test.go:
```diff
@@ -0,0 +1,98 @@
+	"go.etcd.io/bbolt/internal/btesting"
+)
+
+func TestRepeatableRead(t *testing.T) {
```
@serathius Had a great comment that if we want to invest in API surface tests, we should consider the same linearizability library we use in etcd -> i.e. verify that there is linear order between reads and the interleaved writes.

ahrtr replied:
Good point. Let me do this in followup PR.

ahrtr commented:
Thanks @ptabor for the comments, which basically all make sense. I refactored the test case, and also added [todo list](https://github.com/ahrtr/bbolt/blob/4e6e34c5dde63350670a856ef1a9a93821189848/concurrent_test.go#L83-L91), to be resolved in followup PRs,

```
	// TODO (ahrtr):
	//   1. enhance the concurrent test to support more scenarios, e.g. more
	//      RO transactions, random read/write interval, etc.
	//   2. intentionally inject a random failpoint.
	//   3. validate the linearizablity: each reading transaction
	//      should read the value written by previous writing transaction.
	//   4. backup the db file on any errors. Note that the btesting package
	//      also backup the db file when MustCheck fails, so we need to
	//      provide a more generic & integrated solution.
```

ahrtr commented:
cc @ptabor PTAL

ptabor commented on concurrent_test.go:
```diff
@@ -0,0 +1,325 @@
+		minWriteInterval: minWriteInterval,
+		maxWriteInterval: maxWriteInterval,
+
+		errCh:  errCh,
```
not blocking: 
  I would probably pass the operational data to writer.run() directly. 

```
g.Go(func() { return writer.run(ctx, t)}) 
```
or
```
g.Go(writer.runCallback(ctx, t))
```
  - ctx replaces stopCh. 
  - errCh not needed as function called by `Go` can return error and automatically cancel context.

ahrtr replied:
> errCh not needed as function called by Go can return error and automatically cancel context.

We need to stop the test when any worker (including reader and writer) runs into error, see line 153 ~ 156.

```
	select {
	case <-time.After(testDuration):
	case <-errCh:
	}
```

Others comments seem not a big deal.

ptabor commented on concurrent_test.go:
```diff
@@ -0,0 +1,325 @@
+func testResultsDirectory(t *testing.T) string {
+	resultsDirectory, ok := os.LookupEnv("RESULTS_DIR")
+	if !ok {
+		resultsDirectory = "/tmp/"
```
`os.MkdirTemp` ? https://pkg.go.dev/os#MkdirTemp

ahrtr replied:
I just copied the code from etcd :)

serathius replied:
MkdirTemp creates a unique directory name by assign a random suffix to name. For robustness tests I preferred to have predictable file locations to be able to provide stable links in documentation and be able to quickly refresh browser set on same path. Downside is that results are overridden by run.

I think it's up to you to pick

ahrtr replied:
Thanks both @ptabor and  @serathius . 

> Downside is that results are overridden by run

Updated. Let's use `os.MkdirTemp` to avoid this.

ahrtr commented:
Note: I will continue to enhance the test case in followup PRs.
